cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll

```diff
@@ -0,0 +1,50 @@
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+private string strtol() { result = ["strtod", "strtof", "strtol", "strtoll", "strtoq", "strtoul"] }
+
+/**
+ * The standard function `strtol` and its assorted variants
+ */
+private class Strtol extends AliasFunction, ArrayFunction, TaintFunction, SideEffectFunction {
+  Strtol() { this.hasGlobalOrStdOrBslName(strtol()) }
+
+  override predicate hasArrayInput(int bufParam) {
+    // All the functions given by `strtol()` takes a `const char*` input as the first parameter
+    bufParam = 0
+  }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) { bufParam = 0 }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0)
+      or
+      input.isParameterDeref(0)
+    ) and
+    output.isReturnValue()
+  }
+
+  override predicate parameterNeverEscapes(int i) {
+    this.getParameter(i).getUnspecifiedType() instanceof PointerType
```

Comment:
Parameter `0` does escape into parameter `1`. Although we have no way of writing that.

Reply:
Indeed, we don't have a way to specify this, no. I can write it as a comment, though 👍

Reply:
Fixed in 866c0039a7af0ead9d2ecce665c8a468a0f32731. I also simplified the code a bit in the process.